contributors/design-proposals/node/node-fence.md

```diff
@@ -0,0 +1,528 @@
+# Node fencing
+Status: Pending
+
+Version: Alpha
+
+Implementation Owner: Yaniv Bronhaim (ybronhei@redhat.com)
+
+## Overview
+Rationale and architecture for the addition of a fencing mechanism in Kubernetes clusters.
+
+### Motivation
+In Kubernetes a cluster can get into a network partition state between nodes to the api server running on the master node (e.g. network device failure). When that happens their nodes’ status is changed to “not ready” by the node controller and the scheduler cannot know the status of pods running on that node without reaching the nodes’ kubelet service. From such scenario, k8s (since 1.8) defines an eviction timeout (forced by the node controller) such that after 5 minutes pods will enter a termination state in the api server and will try to shutdown gracefully when possible (this happens only if the connectivity returns).
+
+If the pod belongs to a ReplicaSet, once the termination state is set, the ReplicaSet controller will immediately start another instance of that pod.  However when the pod is part of a StatefulSet, the StatefulSet controller won’t start new instance of that pod until the kubelet responds with the pod’s status. This is because there is a permanent association between a StatefulSet member and its storage.  Not waiting would potentially result in multiple copies of a member, all writing to the same volume (leading to corruption or worse).
+
+### Cloud deployments
+When k8s is deployed on a cloud such as AWS or GCE, the autoscaler uses the Cloud Provider APIs to remove failed nodes (Failure detection and allowing recovery of StatefulSets is more of a side effect of the autoscaler which is more focused on using the Cloud Provider APIs to monitor load and trigger scale up/down events), effectively bounding the amount of time the node will stay in the “not ready” state and how long the scheduler will need to wait before it can safely start the pod elsewhere.  However in the case of bare metal, no such mechanism exists and recovery will be blocked until an admin intervenes.
```

Comment:
not all cloud deployments of k8s use the cluster autoscaler.

Reply:
the node controller is actually the one who does this as well.  the node controller removes instances the cloud provided determines as dead.  the node controller cannot remove instances that have kernel soft lockups or other permanent faults that can only be resolved by a restart.

Reply:
However, for the cloud provider, you should at least describe whether a kube add-on is appropriate, or simply the cloud provider's native health checking (i.e. if i have a GCP auto scaling group with an integrated health check, do I need this component?)

Reply:
I couldn't find how to configure node controller to remove instances that the cloud provider signs as unhealthy. can you direct me to this logic?
This paragraph only needs to say that current autoscaler impl doesn't treat partitioned nodes effectively, it waits some time period and then try to scale down the unresponsive nodes. With fencing addon the partitioned nodes can be treated immediately and with more options for actions.
The fence controller logic can use the cloud provider events in addition to the kubelet health check. we also integrate with node-problem-detector that can be used in baremetal to start the fence flow.